src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorUprooted.java

```diff
@@ -0,0 +1,564 @@
+        // 1. we need if reads are lacking or have incomplete quality scores
+        // 2. we add if defaultBaseQualities has a positive value
+        if (defaultBaseQualities < 0) {
+            return read;
+        }
+        byte reads[] = read.getReadBases();
+        byte quals[] = read.getBaseQualities();
+        if (quals == null || quals.length < reads.length) {
+            byte new_quals[] = new byte[reads.length];
+            Arrays.fill(new_quals, defaultBaseQualities);
+            read.setBaseQualities(new_quals);
+        }
+        return read;
+    }
+
+    public RecalibrationTables getTables() {
+        return recalibrationEngine.getRecalibrationTables();
+    }
+
+
+    private ReadTransformer makeReadTransform(){
+        ReadTransformer f0 = BaseRecalibratorUprooted::consolidateCigar;
+
+        ReadTransformer f =
+                f0.andThen(this::setDefaultBaseQualities)
+                .andThen(this::resetOriginalBaseQualities)
+                .andThen(ReadClipper::hardClipAdaptorSequence)
+                .andThen(ReadClipper::hardClipSoftClippedBases);
+
+        return f;
+    }
+
+    @Override
+    public void apply( SAMRecord originalRead, ReferenceContext ref, FeatureContext featureContext ) {
+        throw new RuntimeException("apply Not implemented");
+    }
+
+    /**
+     * For each read at this locus get the various covariate values and increment that location in the map based on
+     * whether or not the base matches the reference at this particular location.
+     * The "features" list doesn't need to be complete but it must cover the read
+     */
+    public void apply( SAMRecord originalRead, final ReferenceContext ref, List<? extends Feature> features ) {
+        ReadTransformer transform = makeReadTransform();
+        final SAMRecord read = transform.apply(originalRead);
+
+        if( ReadUtils.isEmpty(read) ) { return; } // the whole read was inside the adaptor so skip it
+
+        RecalUtils.parsePlatformForRead(read, RAC);
+        if (!RecalUtils.isColorSpaceConsistent(RAC.SOLID_NOCALL_STRATEGY, read)) { // parse the solid color space and check for color no-calls
+            return; // skip this read completely
+        }
+
+        // We've checked in onTraversalStart() that we have a reference, so ref.get() is safe
+        final int[] isSNP = calculateIsSNP(read, ref, originalRead);
+        final int[] isInsertion = calculateIsIndel(read, EventType.BASE_INSERTION);
+        final int[] isDeletion = calculateIsIndel(read, EventType.BASE_DELETION);
+        final int nErrors = nEvents(isSNP, isInsertion, isDeletion);
+
+        // note for efficiency regions we don't compute the BAQ array unless we actually have
+        // some error to marginalize over.  For ILMN data ~85% of reads have no error
+        final byte[] baqArray = nErrors == 0 ? flatBAQArray(read) : calculateBAQArray(read);
+
+        if( baqArray != null ) { // some reads just can't be BAQ'ed
```

Comment:
Please add a comment explaining what's happening here (it's pretty important and somewhat indirect). The snp and indel errors are placed in arrays (zero when not an error of that particular type or not an error). These errors (as stored as offsets from the start of the read) are used by the recalibration engine to update the qualityScoreTable it contains.